lib/backup/azremote/azblob.go

```diff
@@ -41,49 +55,74 @@ type FS struct {
 	Dir string
 
 	client *container.Client
+	env    envLookuper
 }
 
 // Init initializes fs.
 //
 // The returned fs must be stopped when no long needed with MustStop call.
 func (fs *FS) Init() error {
-	if fs.client != nil {
+	switch {
+	case fs.client != nil:
 		logger.Panicf("BUG: fs.Init has been already called")
+	case fs.env == nil:
+		fs.env = envtemplate.LookupEnv
 	}
 
-	for strings.HasPrefix(fs.Dir, "/") {
-		fs.Dir = fs.Dir[1:]
-	}
-	if !strings.HasSuffix(fs.Dir, "/") {
-		fs.Dir += "/"
+	fs.Dir = cleanDirectory(fs.Dir)
+
+	domain := "blob.core.windows.net"
+	if storageDomain, ok := fs.env(envStorageDomain); ok {
+		logger.Infof("Overriding default Azure blob domain with %q", storageDomain)
+		domain = storageDomain
 	}
 
+	connString, hasConnString := fs.env(envStorageAccCs)
+	accountName, hasAccountName := fs.env(envStorageAcctName)
+	accountKey, hasAccountKey := fs.env(envStorageAccKey)
+	useDefault, _ := fs.env(envStorageDefault)
+
 	var sc *service.Client
 	var err error
-	if cs, ok := envtemplate.LookupEnv(envStorageAccCs); ok {
-		sc, err = service.NewClientFromConnectionString(cs, nil)
+
+	// not used if connection string is set
+	serviceURL := fmt.Sprintf("https://%s.%s/", accountName, domain)
+
+	switch {
+	// can't specify any combination of more than one credential
+	case moreThanOne(hasConnString, (hasAccountName && hasAccountKey), (useDefault == "true" && hasAccountName)):
+		logger.Errorf("only one of connection string, account name and key, or default credential can be specified")
+		return errNoCredentials
+	case hasConnString:
+		logger.Infof("Creating AZBlob service client from connection string")
+		sc, err = service.NewClientFromConnectionString(connString, nil)
 		if err != nil {
-			return fmt.Errorf("failed to create AZBlob service client from connection string: %w", err)
+			return fmt.Errorf("failed to create AZBlob service client: %w", err)
```

Comment:
can we put context about used cred types in the error message? If we do, we can remove the Info line above. This is related to the cases below as well. Or am I missing something?

Reply:
As a consumer of the functionality I like seeing what path it takes to generate a connection to azure - should it be Debug level?

Not opposed to wrapping the underlying error with context if you think we should do that as well.

Reply:
Understood. Thanks